### PR TITLE
[bundle] Automatically trampoline to postit

### PR DESCRIPTION
**Pull request by segiddins**
- [x] Specs

~~Except on bundle exec, since that would be too slow~~

~~If you want, I guess we could vendor postit and manually setup the `LOAD_PATH` and `-r` the full path to the vendored postit exe. Your call.~~ Done this

\c @indirect

### CODE REVIEW

**Comment by indirect** [comment]
I thought we already trampolined to the version in the lockfile? (I'm thinking maybe we should merge postit in and treat it as an internal library rather than a vendored external library?)

**Comment by segiddins** [comment]
@indirect postit is what figures out what version is in the lockfile

**Comment by segiddins** [comment]
This has now been completely rewritten, and only really needs extensive testing

\c @RochesterinNYC

**Comment by RochesterinNYC** [comment]
I can take a look at adding specs.

**Comment by segiddins** [comment]
no no you don't have to, I can write specs :P
just saying you might want to review it

File: lib/bundler/cli.rb [hunk 1]
```diff
@@ -199,6 +199,8 @@ def install
       "Force downloading every gem."
     method_option "ruby", :type => :boolean, :banner =>
       "Update ruby specified in Gemfile.lock"
+    method_option "bundler", :type => :string, :lazy_default => "> 0.a", :banner =>
```
Comment (indirect) [inline lib/bundler/cli.rb]: Not sure about defaulting people into a prerelease without them asking for it. Do you feel like that's okay to do?

**Comment by indirect** [comment]
Just to make sure I'm understanding this... the explicit `bundle update --bundler` flag is so that you can short-circuit the trampoline? Some usage docs might be good here 😅

I was initially expecting that it would be possible to manage Bundler's version using the Gemfile and `gem "bundler"`, but of course as I'm typing this I realize that would break backwards compatibility :sob:.

The code looks much less horrible than I was expecting, so great job there! 😁

File: lib/bundler/cli.rb (hunk 1 above)
Comment (segiddins) [inline lib/bundler/cli.rb]: It'll only ever pin them to a pre-release if that's the running version of bundler (i.e. if they've installed a pre-release). Remember that RubyGems' default is to run rereleases for binaries

Comment (indirect) [inline lib/bundler/cli.rb]: Oh yeah! Good point.

**Comment by segiddins** [comment]
> Just to make sure I'm understanding this... the explicit bundle update --bundler flag is so that you can short-circuit the trampoline? Some usage docs might be good here 😅

It something that's easy to check for, and is a simple way to force `Definition#to_lock` to include the new bundler version :P

**Comment by indirect** [comment]
@homu r+

**Comment by segiddins** [comment]
@indirect dear God please don't merge this without tests

**Comment by indirect** [comment]
lol whooooops

File: lib/bundler/shared_helpers.rb [hunk 2]
```diff
@@ -169,6 +169,7 @@ def set_bundle_variables
 
       # Set BUNDLE_GEMFILE
       ENV["BUNDLE_GEMFILE"] = find_gemfile.to_s
+      ENV["BUNDLER_VERSION"] = Bundler::VERSION
```
Comment (RochesterinNYC) [inline lib/bundler/shared_helpers.rb]: Should this also be an `if defined?(Bundler::VERSION)` kind of case?

File: lib/bundler/version.rb [hunk 3]
```diff
@@ -7,5 +7,5 @@ module Bundler
   # We're doing this because we might write tests that deal
   # with other versions of bundler and we are unsure how to
   # handle this better.
-  VERSION = "1.12.5" unless defined?(::Bundler::VERSION)
+  VERSION = "1.12.9" unless defined?(::Bundler::VERSION)
```
Comment (RochesterinNYC) [inline lib/bundler/version.rb]: What's the reason for this version bump?

Comment (segiddins) [inline lib/bundler/version.rb]: 😁testing. it shouldn't be there

File: lib/bundler/shared_helpers.rb (hunk 2 above)
Comment (segiddins) [inline lib/bundler/shared_helpers.rb]: nah this is inside the bundler module

**Comment by segiddins** [comment]
Basic specs added, what else would you like to see?

File: spec/other/trampoline_spec.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+require "spec_helper"
+
+describe "bundler version trampolining" do
+  before do
+    # ENV["BUNDLE_DISABLE_POSTIT"] = nil
```
Comment (indirect) [inline spec/other/trampoline_spec.rb]: is this for debugging?

**Comment by indirect** [comment]
I'd like to see a test around the error message if postit's automatic installation fails, but otherwise this is looking pretty good!

File: spec/other/trampoline_spec.rb (hunk 4 above)
Comment (segiddins) [inline spec/other/trampoline_spec.rb]: 😭 yes. I'm tired

File: spec/other/trampoline_spec.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    before do
+      simulate_bundler_version("1.11.1") do
+        install_gemfile ""
+      end
+    end
+
+    it "updates to the specified version" do
+      # HACK: since no released bundler version actually supports this feature!
+      bundle "update --bundler=1.12.0", :expect_err => true
+      expect(out).to include("Unknown switches '--bundler=1.12.0'")
```
Comment (indirect) [inline spec/other/trampoline_spec.rb]: Shouldn't `bundle update --bundler=1.12.0` at least lock to Bundler 1.12.0, and then automatically install it and trampoline to it on the next execution?

**Comment by segiddins** [comment]
@indirect done

File: spec/other/trampoline_spec.rb (hunk 5 above)
Comment (segiddins) [inline spec/other/trampoline_spec.rb]: yes, but I couldn't figure out how to do that without rewriting how all of definition locking works and I'm lazy and it doesn't seem worth it

Comment (indirect) [inline spec/other/trampoline_spec.rb]: Does this test exist to verify that specified versions are not supported? I guess I'm confused by the spec string of "it updates to the specified version" combined with the expectation of "Unknown switch".

Comment (segiddins) [inline spec/other/trampoline_spec.rb]: I'd update it as soon as we have a released version that actually accepts the `--bundle` option

**Comment by segiddins** [comment]
@indirect this should now be green 🎆

**Comment by indirect** [comment]
:tada:

@homu r+

**Comment by indirect** [comment]
@homu r+ one more time

**Pull request merged**